File: sph/sph.hs
{-
https://github.com/takagi/blog-codes/blob/master/20091101/haskell/sph.hs

ToDo:
  □【重要】近傍マップは領域より一回り大きくとる。
    粒子がさらにその外に来た場合にはエラーとする。
  □subversionで管理
  □近傍マップをcalc_amountとadvanceとで２回作るのがヤダ
  □だんだん１ステップの計算速度が遅くなっていないか？
  □cppと比べてどれくらい遅いか？
    □cppでかかる時間は？
    □hsでかかる時間は？
  □ユニットテスト
  □OpenGL
  □マーチングキューブ
  □関数の整理
  □プロファイラを使ってチューニング
  □近傍マップ関連のデータ型を整理する
  ■boundary_condition
  ■output
  ■massの導入
    - Particleの定義にmassを入れる
    - Particleの初期化にmassを入れる
    - advanceで新しいParticleを作るときにmassを入れる
    - 速度を力から計算する（つまり、aではなくfから作る）
    - など
  ■複数粒子の導入
    - 初期状態の型を[Particle]にする
    - simulationの型を [Particle] -> [[Particle]] にする
    - advanceの型を [Particle] -> [Particle] にする
  ■初期配置->もっと粒子数が少なくて済むようにする
  ■近傍マップ
  ■sph_hはsimscaleをかけた後のもの？
    →その前提で書いている
  ■位置や速度がNaNになってしまう問題の解決
    →dtが大きすぎて、近傍マップの範囲外に粒子が移動してしまっていたから
  ×massは不要
  ■近傍マップの作成が遅すぎる
-}

import Data.Vector
import Data.Array
import System

{-
  Constants
-}
dt          = 0.004
simscale    = 0.004
sph_h       = 0.01
visc        = 0.2
restdensity = 600.0
intstiff    = 3.0
pmass       = 0.00020543
pdist       = (pmass / restdensity) ** (1/3)
e           = 2.71828183
radius      = 0.004
extstiff    = 10000.0
extdamp     = 256.0
epsilon     = 0.00001
limit       = 200.0
g           = Vector3 0 (-9.8) 0
box_min     = Vector3 0  0 (-10)
--box_max     = Vector3 300 50 10
box_max     = Vector3 20 50 10
init_min    = Vector3 0  0 (-10)
init_max    = Vector3 10 20 10

main = do n <- getArgs
          putNumberOf ps0
          output $ take (read (n!!0)) $ simulation ps0

-- | 粒子の定義
data Particle =
  Particle { pos, vel :: {-# UNPACK #-} !Vector3,
             mass, rho, prs :: {-# UNPACK #-} !Scalar }
  deriving (Show, Eq)

ps0 :: [Particle]
ps0 = [ Particle
        { pos  = Vector3 x y z,
          vel  = Vector3 0 0 0,
          mass = pmass,
          rho  = 0,
          prs  = 0 }
      | x <- xs, y <- ys, z <- zs ]
  where
    d  = pdist / simscale * 0.95
    xs = let minx = v3x init_min
             maxx = v3x init_max
         in [minx+d, minx+d+d .. maxx-d]
    ys = let miny = v3y init_min
             maxy = v3y init_max
         in [miny+d, miny+d+d .. maxy-d]
    zs = let minz = v3z init_min
             maxz = v3z init_max
         in [minz+d, minz+d+d .. maxz-d]
{-
ps0 = [ Particle { pos = initial_position x y z,
                   vel = (simscale*30.0) *<> (initial_velocity x y z),
                   mass = pmass,
                   rho  = 0,
                   prs  = 0  }
        | x <- xs, y <- ys, z <- zs, is_under_surface x y z ]
  where d  = pdist / simscale * 0.95
        xs = let minx = v3x box_min
                 maxx = v3x box_max
             in [minx+d, minx+d+d .. maxx]
        ys = let miny = v3y box_min
                 maxy = v3y box_max
             in [miny+d, miny+d+d .. maxy]
        zs = let minz = v3z box_min
                 maxz = v3z box_max
             in [minz+d, minz+d+d .. maxz]
        is_under_surface x y _
          | y > d+h*_sech**2 = False
          | x <= 50          = y >= 0.0
          | x <= 150         = -x+100/18*y+50 >= 0.0
          | otherwise        = if y < 18 then False else True
          where _sech = sech((sqrt((3*h)/(4*d**3)))*(x-20))
                h     = 7
                d     = 10
                g     = 9.8
        initial_position x y z = Vector3 x y z
        initial_velocity x y _ =
            let _x = (sqrt(g*d)) * h/d * _sech**2
                _y = (sqrt(g*d)) * (h/d)**1.5 * y/d * _sech**2 * _tanh
                h  = 7
                d  = 10
                g  = 9.8
                _sech = sech((sqrt((3*h)/(4*d**3)))*(x-20))
                _tanh = Main.tanh((sqrt((3*h)/(4*d**3)))*(x-20))
            in Vector3 _x _y 0

sech :: Scalar -> Scalar
sech x = 2.0 / (e**x + e**(-x))

tanh :: Scalar -> Scalar
tanh x = (e**x - e**(-x)) / (e**x + e**(-x))
-}

-- | NeighborMap の定義
type NeighborIdx = (Int,Int,Int)
type NeighborMap = Array NeighborIdx [Particle]
type NeighborList = [(NeighborIdx,[Particle])]

neighbor_idx :: Vector3 -> NeighborIdx
neighbor_idx (Vector3 x y z) =
  (toIx x, toIx y, toIx z)
  where
    toIx :: Scalar -> Int
    toIx x = floor(x*simscale/sph_h)

neighbor_map :: [Particle] -> NeighborMap
neighbor_map ps =
  let ix_min   = let (ix,iy,iz) = neighbor_idx box_min
                 in (ix-1,iy-1,iz-1)
      ix_max   = let (ix,iy,iz) = neighbor_idx box_max
                 in (ix+1,iy+1,iz+1)
      nbr_list = _neighbor_list ps []
  in array (ix_min,ix_max)
     [ (ix,case lookup ix nbr_list of
             Just x  -> x
             Nothing -> [])
     | ix <- range (ix_min,ix_max) ]
  where
    _neighbor_list :: [Particle] -> NeighborList -> NeighborList
    _neighbor_list [] lst = lst
    _neighbor_list (p:ps) lst =
      insert_neighbor p (_neighbor_list ps lst)

insert_neighbor :: Particle -> NeighborList -> NeighborList
insert_neighbor p@(Particle {pos=r}) nbr_map =
  let ixp = neighbor_idx r
  in _insert_neighbor ixp p nbr_map
  where
    _insert_neighbor :: NeighborIdx -> Particle -> NeighborList
                     -> NeighborList
    _insert_neighbor ixp p (x@(ixx,ps):xs)
      | ixp == ixx = (ixx,(p:ps)) : xs
      | otherwise  = x : _insert_neighbor ixp p xs
    _insert_neighbor ixp p [] = [(ixp,[p])]

neighbor :: NeighborMap -> Vector3 -> [Particle]
neighbor nbr_map r =
  let (_ix,_iy,_iz) = neighbor_idx r
      ix_min   = let (ix,iy,iz) = neighbor_idx box_min
                 in (ix-1,iy-1,iz-1)
      ix_max   = let (ix,iy,iz) = neighbor_idx box_max
                 in (ix+1,iy+1,iz+1)
  in flatten [ nbr_map!(ix,iy,iz)
               | ix <- [_ix-1, _ix, _ix+1]
               , iy <- [_iy-1, _iy, _iy+1]
               , iz <- [_iz-1, _iz, _iz+1]
               , ix_min <= (ix,iy,iz)
               , (ix,iy,iz) <= ix_max ]

-- | シミュレーション用の関数
simulation :: [Particle] -> [[Particle]]
simulation ps =
  let ps'  = calc_amount ps $ neighbor_map ps
      ps'' = advance ps' $ neighbor_map ps'
  in ps : simulation ps''

calc_amount :: [Particle] -> NeighborMap -> [Particle]
calc_amount ps nbr_map = map _calc_amount ps
  where
    _calc_amount p@(Particle r v m _ _) =
      let rho = density (neighbor nbr_map r) p
          prs = (rho - restdensity ) * intstiff
      in Particle r v m rho prs

advance :: [Particle] -> NeighborMap -> [Particle]
advance ps nbr_map = map _advance ps
  where
    _advance p@(Particle r v m rho prs) =
      let r_next = r + v_next * dt / fromScalar(simscale)
          v_next = v + a * dt
          a      = g + (boundary_condition p _a)
          _a     = f / fromScalar(rho)
          f      = force (neighbor nbr_map r) p
      in Particle r_next v_next m rho prs

density :: [Particle] -> Particle -> Scalar
density ps i = sum $ map exp ps
  where
    exp :: Particle -> Scalar
    exp j = let Particle { pos = r_i } = i
                Particle { pos = r_j, mass = m_j } = j
                dr = simscale *<> (r_i - r_j)
            in m_j * (poly6_kernel dr)

force :: [Particle] -> Particle -> Vector3
force ps p =
  pressure_term ps p + viscosity_term ps p

pressure_term :: [Particle] -> Particle -> Vector3
pressure_term ps i =
  -(sum $ map exp $ filter (/=i) ps)
  where
    exp :: Particle -> Vector3
    exp j = let Particle { pos = r_i, prs = p_i    } = i
                Particle { pos = r_j, prs = p_j,
                           mass = m_j, rho = rho_j } = j
                dr = simscale *<> (r_i - r_j)
            in (m_j * (p_i + p_j) / (2.0*rho_j))
                                    *<> (grad_spiky_kernel dr)

viscosity_term :: [Particle] -> Particle -> Vector3
viscosity_term ps i =
  visc * (sum $ map exp $ filter (/=i) ps)
  where
    exp :: Particle -> Vector3
    exp j = let Particle { pos = r_i, vel = v_i    } = i
                Particle { pos = r_j, vel = v_j,
                           mass = m_j, rho = rho_j } = j
                dr = simscale *<> (r_i - r_j)
            in (m_j / rho_j * (rap_visc_kernel dr)) *<> (v_j - v_i)

boundary_condition :: Particle -> Vector3 -> Vector3
boundary_condition p@Particle {pos=r,vel=v} a =
  x_wall_max $ x_wall_min
  $ y_wall_max $ y_wall_min
  $ z_wall_max $ z_wall_min
  $ accel_limit a
  where
    x_wall_max = wall ((v3x box_max)-(v3x r)) (Vector3 (-1) 0 0)
    x_wall_min = wall ((v3x r)-(v3x box_min)) (Vector3 1 0 0)
    y_wall_max = wall ((v3y box_max)-(v3y r)) (Vector3 0 (-1) 0)
    y_wall_min = wall ((v3y r)-(v3y box_min)) (Vector3 0 1 0)
    z_wall_max = wall ((v3z box_max)-(v3z r)) (Vector3 0 0 (-1))
    z_wall_min = wall ((v3z r)-(v3z box_min)) (Vector3 0 0 1)
    wall d norm a = let diff = 2.0*radius - d*simscale
                        adj  = extstiff*diff-extdamp*(vdot norm v)
                    in if diff > epsilon
                      then a + (adj *<> norm)
                      else a
    accel_limit a = let speed  = vmag a
                    in if speed > limit
                      then (limit / speed) *<> a
                      else a

{-
  Kernel functions
-}

poly6_kernel :: Vector3 -> Scalar
poly6_kernel x
  | r <= sph_h = 315.0 / (64.0*pi*(sph_h**9)) * (sph_h**2-r**2)**3
  | otherwise  = 0
  where
    r = vmag x

grad_spiky_kernel :: Vector3 -> Vector3
grad_spiky_kernel x
  | r <= sph_h = ((-45.0) / (pi*(sph_h**6)) * ((sph_h-r)**2) / r) *<> x
  | otherwise  = Vector3 0 0 0
  where
    r = vmag x

rap_visc_kernel :: Vector3 -> Scalar
rap_visc_kernel x
  | r <= sph_h = 45.0 / (pi*(sph_h**6)) * (sph_h-r)
  | otherwise  = 0
  where
    r = vmag x


{-
  Utilities
-}

flatten :: [[a]] -> [a]
flatten = foldr (++) []


{-
  Actions and functions for output
-}

putNumberOf :: Show a => [a] -> IO()
putNumberOf = putStrLn . show . length

output :: [[Particle]] -> IO()
output pss = _output pss 0
  where
    _output :: [[Particle]] -> Int -> IO()
    _output (ps:pss) i = do output_povs ps i
                            --output_csvs ps i
                            _output pss (i+1)
    _output []       _ = return ()

output_povs :: [Particle] -> Int -> IO()
output_povs ps i =
  do putStrLn ("processing "++ (file_name i)++" ...")
     write ps i
  where
    write :: [Particle] -> Int -> IO()
    write ps i = writeFile (file_name i) (body ps)

    file_name :: Int -> String
    file_name i = "result" ++ (rjust8 i) ++ ".pov"

    body :: [Particle] -> String
    body ps = "#include \"colors.inc\"\n"
           ++ "camera {\n"
           ++ "  location <10, 30, -40.0>\n"
           ++ "  look_at <10, 10, 0.0>\n"
           ++ "}\n"
           ++ "light_source { <0, 30, -30> color White }\n"
           ++ (flatten $ map sphere ps)

    sphere :: Particle -> String
    sphere p = let Vector3 x y z = pos p
               in "sphere {\n"
               ++ "  <"++(show x)++","++(show y)++","++(show z)
                                                        ++">,0.5\n"
               ++ "  texture {\n"
               ++ "    pigment { color Yellow }\n"
               ++ "  }\n"
               ++ "}\n";

output_csvs :: [Particle] -> Int -> IO()
output_csvs ps i =
  do putStrLn ("processing "++ (file_name i)++" ...")
     write ps i
  where
    write :: [Particle] -> Int -> IO()
    write ps i = writeFile (file_name i) (header++body ps)

    file_name :: Int -> String
    file_name i = "result" ++ (rjust8 i) ++ ".csv"

    header :: String
    header = "x,y,z,vx,vy,vz,rho,prs\n"

    body :: [Particle] -> String
    body ps = flatten $ map line ps

    line :: Particle -> String
    line Particle {pos = (Vector3 x y z),
                   vel = (Vector3 vx vy vz),
                   rho = rho,
                   prs = prs}
         = (flatten $ map (\x -> show x ++ ",")
                          [x,y,z,vx,vy,vz,rho,prs]) ++ "\n"

rjust :: Int -> Int -> String
rjust width s =
  replicate (width - (length $ show s)) '0' ++ (show s)
rjust8 =
  rjust 8

File: sph/sph.cpp
// g++ -O2 -o sph sph.cpp ; ./sph 300
// https://github.com/takagi/blog-codes/blob/master/20091101/cpp/sph.cpp
#include <vector>
#include <list>
#include <map>
#include <iostream>
#include <fstream>
#include <cmath>

/*----------------------------
  Declaration of vec3
----------------------------*/

struct vec3 {
  double x, y, z;
  vec3() {
    x=0; y=0; z=0;
  }
  vec3( const double _x, const double _y, const double _z ) {
    x=_x; y=_y;z=_z;
  }
};

inline void operator+= ( vec3 &v, const vec3 a ){
  v.x += a.x; v.y += a.y; v.z += a.z;
}

inline void operator*= ( vec3 &v, const double a ){
  v.x *= a; v.y *= a; v.z *= a;
}

inline bool operator== ( const vec3 a, const vec3 b )
{
  return a.x==b.x && a.y==b.y && a.z==b.z;
}

inline vec3 operator- ( const vec3 a )
{
  return vec3( -a.x, -a.y, -a.z);
}

inline vec3 operator+ ( const vec3 a, const vec3 b )
{
  vec3 v;
  v.x = a.x + b.x;
  v.y = a.y + b.y;
  v.z = a.z + b.z;
  return v;
}

inline vec3 operator- ( const vec3 a, const vec3 b )
{
  vec3 v;
  v.x = a.x - b.x;
  v.y = a.y - b.y;
  v.z = a.z - b.z;
  return v;
}

inline vec3 operator* ( const double a, const vec3 b )
{
  vec3 v;
  v.x = a * b.x;
  v.y = a * b.y;
  v.z = a * b.z;
  return v;
}

inline vec3 operator* ( const vec3 a, const double b )
{
  return b * a;
}

inline vec3 operator/ ( const vec3 a, const double b )
{
  if ( b != 0 )
    return a * ( 1 / b );
  else
    {
      std::cout << "error: division by 0." << std::endl;
      exit(1);
    }
}

inline double dot ( const vec3 a, const vec3 b )
{
  return a.x*b.x + a.y*b.y + a.z*b.z;
}

inline double mag2 ( const vec3 a )
{
  return a.x*a.x+a.y*a.y+a.z*a.z;
}

inline double mag ( const vec3 a )
{
  return sqrt(mag2(a));
}


/*----------------------------
  Constants
----------------------------*/

static const double SPH_RESTDENSITY = 600.0;
static const double SPH_INTSTIFF    = 3.0;
static const double SPH_PMASS       = 0.00020543;
static const double SPH_SIMSCALE    = 0.004;
static const double H               = 0.01;
static const double PI              = 3.141592653589793;
static const double DT              = 0.004;
static const double SPH_VISC        = 0.2;
static const double SPH_LIMIT       = 200.0;
static const double SPH_RADIUS      = 0.004;
static const double SPH_EPSILON     = 0.00001;
static const double SPH_EXTSTIFF    = 10000.0;
static const double SPH_EXTDAMP     = 256.0;
static const double SPH_PDIST       = pow(SPH_PMASS / SPH_RESTDENSITY,
                                          1.0/3.0 );
static const vec3   MIN             = vec3(  0.0,  0.0, -10.0 );
static const vec3   MAX             = vec3( 20.0, 50.0,  10.0 );
static const vec3   INIT_MIN        = vec3(  0.0,  0.0, -10.0 );
static const vec3   INIT_MAX        = vec3( 10.0, 20.0,  10.0 );
static const double Poly6Kern       = 315.0 / ( 64.0 * PI * pow( H, 9 ) );
static const double SpikyKern       = -45.0 / ( PI * pow( H, 6 ) );
static const double LapKern         = 45.0 / ( PI * pow( H, 6 ) );


/*----------------------------
  Declaration of Particle
----------------------------*/

struct Particle
{
  vec3 pos, vel, f;
  double rho, prs;
};
typedef std::vector<Particle> Particles;

Particles* new_particles()
{
  Particles* p_ps;
  p_ps = new Particles();

  double d = SPH_PDIST / SPH_SIMSCALE * 0.95;
  for ( double x = INIT_MIN.x+d; x <= INIT_MAX.x-d; x += d )
  for ( double y = INIT_MIN.y+d; y <= INIT_MAX.y-d; y += d )
  for ( double z = INIT_MIN.z+d; z <= INIT_MAX.z-d; z += d )
    {
      Particle p;
      p.pos.x = x; p.pos.y = y; p.pos.z = z;
      p.vel.x = 0.0; p.vel.y = 0.0; p.vel.z = 0.0;
      p.f.x = 0.0; p.f.y = 0.0; p.f.z = 0.0;
      p.rho  = 0.0;
      p.prs  = 0.0;
      p_ps->push_back( p );
    }
  return p_ps;
}

void delete_particles( Particles* p_ps )
{
  delete p_ps;
}

int particles_size( Particles* p_ps )
{
  return p_ps->size();
}

#define FOR_EACH_PARTICLE( p_p, p_ps ) \
  for( Particles::iterator (p_p) = (p_ps)->begin(); (p_p) != (p_ps)->end(); (p_p)++ )


/*----------------------------
  Declaration of NeighborMap
----------------------------*/

typedef long NeighborMapIdx;
typedef std::list< Particle* > ParticlePtrs;
typedef std::map< NeighborMapIdx, ParticlePtrs > NeighborMap;

NeighborMap*   new_neighbor_map( Particles* p_ps );
void           _insert_neighbor_map( Particle*, NeighborMap* );
void           delete_neighbor_map( NeighborMap* );
ParticlePtrs   neighbor( NeighborMap*, vec3 );
NeighborMapIdx neighbor_map_idx( vec3 );

#define FOR_EACH_PARTICLE_PTR( p_ptr, p_ptrs ) \
  for( ParticlePtrs::iterator p_ptr = (p_ptrs)->begin(); (p_ptr) != (p_ptrs)->end(); (p_ptr)++ )

NeighborMap* new_neighbor_map( Particles* p_ps )
{
  NeighborMap* p_nbr_map;
  p_nbr_map = new NeighborMap();
  FOR_EACH_PARTICLE( p_p, p_ps )
    _insert_neighbor_map( &*p_p, p_nbr_map );
  return p_nbr_map;
}

void _insert_neighbor_map( Particle* p_p, NeighborMap* p_nbr_map )
{
  NeighborMapIdx ix = neighbor_map_idx( p_p->pos );
  NeighborMap::iterator iter = p_nbr_map->find( ix );
  if ( iter != p_nbr_map->end() )
    {
      iter->second.push_back( p_p );
    }
  else
    {
      ParticlePtrs ptrs;
      ptrs.push_back( p_p );
      p_nbr_map->insert( NeighborMap::value_type( ix, ptrs ) );
    }
}

void delete_neighbor_map( NeighborMap* p_nbr_map )
{
  delete p_nbr_map;
}

ParticlePtrs neighbor( NeighborMap* p_nbr_map, vec3 r )
{
  ParticlePtrs ptrs;
  double d = H / SPH_SIMSCALE;
  for ( int x=-1; x<2; x++ )
  for ( int y=-1; y<2; y++ )
  for ( int z=-1; z<2; z++ )
    {
      vec3 v( r.x+(double)x*d, r.y+(double)y*d, r.z+(double)z*d );
      if ( MIN.x <= v.x && v.x <= MAX.x &&
           MIN.y <= v.y && v.y <= MAX.y &&
           MIN.z <= v.z && v.z <= MAX.z )
        {
          NeighborMapIdx ix = neighbor_map_idx( v );
          NeighborMap::iterator x = p_nbr_map->find(ix);
          if ( x != p_nbr_map->end() )
            {
              FOR_EACH_PARTICLE_PTR( p_ptr, &(x->second) )
                {
                  ptrs.push_back( *p_ptr );
                }
            }
        }
    }
  return ptrs;
}

NeighborMapIdx neighbor_map_idx( vec3 r )
{
  int x, y, z;
  int mx, my;
  double d;
  d  = H / SPH_SIMSCALE;
  x  = floor( (r.x - MIN.x) / d );
  y  = floor( (r.y - MIN.y) / d );
  z  = floor( (r.z - MIN.z) / d );
  mx = floor( (MAX.x - MIN.x) / d );
  my = floor( (MAX.y - MIN.y) / d );
  return x + y * mx + z * mx * my;
}


/*----------------------------
  Functions for simulation
----------------------------*/

void simulation( Particles* );
void calc_amount( Particles*, NeighborMap* );
void calc_force( Particles*, NeighborMap* );
void advance( Particles*, NeighborMap* );

void simulation( Particles* p_ps )
{
  NeighborMap* p_nbr_map;
  p_nbr_map = new_neighbor_map( p_ps );
  calc_amount( p_ps, p_nbr_map );
  calc_force( p_ps, p_nbr_map );
  advance( p_ps, p_nbr_map );
  delete_neighbor_map( p_nbr_map );
}

void calc_amount( Particles* p_ps, NeighborMap* p_nbr_map )
{
  double H2, sum, r2, c;
  vec3 dr;
  ParticlePtrs ptrs;
  Particle* p_pj;

  H2 = H*H;

  FOR_EACH_PARTICLE( p_p, p_ps )
    {
      sum  = 0.0;
      ptrs = neighbor( p_nbr_map, p_p->pos );
      FOR_EACH_PARTICLE_PTR( p_ptr, &ptrs )
        {
          p_pj= *p_ptr;
          dr = (p_p->pos - p_pj->pos) * SPH_SIMSCALE;
          r2 = mag2( dr );
          if ( H2 > r2 )
            {
              c = H2 - r2;
              sum += c * c * c;
            }
        }
      p_p->rho = sum * SPH_PMASS * Poly6Kern;
      p_p->prs = ( p_p->rho - SPH_RESTDENSITY ) * SPH_INTSTIFF;
      p_p->rho = 1.0 / p_p->rho;
    }
}

void calc_force( Particles* p_ps, NeighborMap* p_nbr_map )
{
  double pterm, vterm, r, c;
  vec3 dr, force, fcurr;
  ParticlePtrs ptrs;
  Particle* p_pj;

  FOR_EACH_PARTICLE( p_p, p_ps )
    {
      force = vec3( 0.0, 0.0, 0.0 );
      ptrs = neighbor( p_nbr_map, p_p->pos );
      FOR_EACH_PARTICLE_PTR( p_ptr, &ptrs )
        {
          p_pj = *p_ptr;
          if ( p_p->pos == p_pj->pos ) continue;
          dr = (p_p->pos - p_pj->pos) * SPH_SIMSCALE;
          r  = mag( dr );
          if ( H > r )
            {
              c = H - r;
              pterm = -0.5 * c * SpikyKern * (p_p->prs + p_pj->prs) / r;
              vterm = LapKern * SPH_VISC;
              fcurr = pterm * dr + vterm * (p_pj->vel - p_p->vel);
              fcurr *= c * p_p->rho * p_pj->rho;
              force += fcurr;
            }
        }
      p_p->f = force;
    }
}

void advance( Particles* p_ps, NeighborMap* p_nbr_map )
{
  vec3 accel, g, norm;
  double speed, diff, adj;

  g = vec3( 0.0, -9.8, 0.0 );
  FOR_EACH_PARTICLE( p_p, p_ps )
    {
      accel = p_p->f * SPH_PMASS;

      speed = mag2( accel );
      if ( speed > SPH_LIMIT*SPH_LIMIT ) {
        accel *= SPH_LIMIT / sqrt(speed);
      }

      // Z-axis walls
      diff = 2.0 * SPH_RADIUS - ( p_p->pos.z - MIN.z ) * SPH_SIMSCALE;
      if ( diff > SPH_EPSILON )
        {
          norm = vec3( 0.0, 0.0, 1.0 );
          adj = SPH_EXTSTIFF * diff - SPH_EXTDAMP * dot( norm, p_p->vel );
          accel += adj * norm;
        }
      diff = 2.0 * SPH_RADIUS - ( MAX.z - p_p->pos.z ) * SPH_SIMSCALE;
      if ( diff > SPH_EPSILON )
        {
          norm = vec3( 0.0, 0.0, -1.0 );
          adj = SPH_EXTSTIFF * diff - SPH_EXTDAMP * dot( norm, p_p->vel );
          accel += adj * norm;
        }

      // X-axis walls
      diff = 2.0 * SPH_RADIUS - ( p_p->pos.x - MIN.x ) * SPH_SIMSCALE;
      if ( diff > SPH_EPSILON )
        {
          norm = vec3( 1.0, 0.0, 0.0 );
          adj = SPH_EXTSTIFF * diff - SPH_EXTDAMP * dot( norm, p_p->vel );
          accel += adj * norm;
        }
      diff = 2.0 * SPH_RADIUS - ( MAX.x - p_p->pos.x ) * SPH_SIMSCALE;
      if ( diff > SPH_EPSILON )
        {
          norm = vec3( -1.0, 0.0, 0.0 );
          adj = SPH_EXTSTIFF * diff - SPH_EXTDAMP * dot( norm, p_p->vel );
          accel += adj * norm;
        }

      // Y-axis walls
      diff = 2.0 * SPH_RADIUS - ( p_p->pos.y - MIN.y ) * SPH_SIMSCALE;
      if ( diff > SPH_EPSILON )
        {
          norm = vec3( 0.0, 1.0, 0.0 );
          adj = SPH_EXTSTIFF * diff - SPH_EXTDAMP * dot( norm, p_p->vel );
          accel += adj * norm;
        }
      diff = 2.0 * SPH_RADIUS - ( MAX.y - p_p->pos.y ) * SPH_SIMSCALE;
      if ( diff > SPH_EPSILON )
        {
          norm = vec3( 0.0, -1.0, 0.0 );
          adj = SPH_EXTSTIFF * diff - SPH_EXTDAMP * dot( norm, p_p->vel );
          accel += adj * norm;
        }

      accel += g;
      p_p->vel += accel * DT;
      p_p->pos += p_p->vel * DT / SPH_SIMSCALE;
    }
}


/*----------------------------
  Functions for output
----------------------------*/

void output_particles( Particles* p_ps, int cnt )
{
  char file_name[19];
  sprintf(file_name, "result%08d.pov", cnt);
  std::cout << "processing " << file_name << " ..." << std::endl;
  std::ofstream f( file_name );
  if ( ! f )
    {
      std::cerr << "cannot open " << file_name << std::endl;
      exit(1);
    }

  f << "#include \"colors.inc\"\n"
    << "camera {\n"
    << "  location <10, 30, -40.0>\n"
    << "  look_at <10, 10, 0.0>\n"
    << "}\n"
    << "light_source { <0, 30, -30> color White }\n";
  FOR_EACH_PARTICLE( p_p, p_ps )
  {
    vec3 r = p_p->pos;
    f << "sphere {\n"
      << "  <" << r.x << ", " << r.y << ", " << r.z << ">, 0.5\n"
      << "  texture {\n"
      << "    pigment { color Yellow }\n"
      << "  }\n"
      << "}\n";
  }
  f << std::endl;
}


/*----------------------------
  main function
----------------------------*/

int main( int argc, char** argv )
{
  int n;
  if ( argc != 2 )
    {
      std::cout << "usage: ./sph steps" << std::endl;
      exit(-1);
    }
  else
    n = atoi( argv[1] );

  Particles* p_ps = new_particles();
  std::cout << "# of particles : " << particles_size( p_ps ) << std::endl;
  for ( int i = 0 ; i < n; i++ )
    {
      output_particles( p_ps, i );
      simulation( p_ps );
    }
  delete_particles( p_ps );
}
